caches_bench/caches_bench_test.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"testing"
+	"time"
+	"github.com/allegro/bigcache"
+	"github.com/coocood/freecache"
+	"fmt"
+	"math/rand"
+)
+
+var maxEntrySize = 256
+
+func BenchmarkMapSet(b *testing.B) {
+	m := make(map[string][]byte)
+	for i := 0; i < b.N; i++ {
+		m[key(i)] = value()
+	}
+}
+
+func BenchmarkFreeCacheSet(b *testing.B) {
+	cache := freecache.NewCache(b.N * maxEntrySize)
+	for i := 0; i < b.N; i++ {
+		cache.Set([]byte(key(i)), value(), 0)
+	}
+}
+
+func BenchmarkBigCacheSet(b *testing.B) {
+	cache := initBigCache(b.N)
+	for i := 0; i < b.N; i++ {
+		cache.Set(key(i), value())
+	}
+}
+
+func BenchmarkMapGet(b *testing.B) {
+	b.StopTimer()
+	m := make(map[string][]byte)
+	for i := 0; i < b.N; i++ {
+		m[key(i)] = value()
+	}
+	b.StartTimer()
+	hitCount := 0
+	for i := 0; i < b.N; i++ {
+		if m[key(i)] != nil {
+			hitCount++
+		}
+	}
+}
+
+func BenchmarkFreeCacheGet(b *testing.B) {
+	b.StopTimer()
+	cache := freecache.NewCache(b.N * maxEntrySize)
+	for i := 0; i < b.N; i++ {
+		cache.Set([]byte(key(i)), value(), 0)
+	}
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		cache.Get([]byte(key(i)))
+	}
+}
+
+func BenchmarkBigCacheGet(b *testing.B) {
+	b.StopTimer()
+	cache := initBigCache(b.N)
+	for i := 0; i < b.N; i++ {
+		cache.Set(key(i), value())
+	}
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		cache.Get(key(i))
+	}
+}
+
+func BenchmarkBigCacheSetParallel(b *testing.B) {
+	cache := initBigCache(b.N)
+	rand.Seed(time.Now().Unix())
+
+	b.RunParallel(func(pb *testing.PB) {
+		id := rand.Intn(1000)
+		counter := 0
+		for pb.Next() {
+			cache.Set(parallelKey(id, counter), value())
+			counter = counter + 1
+		}
+	})
+}
+
+func BenchmarkFreeCacheSetParallel(b *testing.B) {
+	cache := freecache.NewCache(b.N * maxEntrySize)
+	rand.Seed(time.Now().Unix())
+
+	b.RunParallel(func(pb *testing.PB) {
+		id := rand.Intn(1000)
+		counter := 0
+		for pb.Next() {
+			cache.Set([]byte(parallelKey(id, counter)), value(), 0)
+			counter = counter + 1
+		}
+	})
+}
```

Comment:
I'd like to see test with eviction on to compare how cache behave when it adds, deletes and search for entries at same time.